File: src/AntiPilot/Log.cs
namespace AntiPilot;

/// <summary>
/// Tiny append-only log. The key-press path runs with no UI, so this is the only
/// way to find out afterwards why nothing happened.
/// </summary>
internal static class Log
{
    private static readonly object Gate = new();

    /// <summary>
    /// Serialises writes across processes, not just threads.
    ///
    /// Several AntiPilot processes are alive at once by design — the tray, the settings window, one
    /// per key press, and two at a time for every double press — and appends from different
    /// processes interleave inside a line, which was observed shredding entries in exactly the
    /// double-press case. The in-process lock below cannot see any of that.
    /// </summary>
    private static readonly Mutex FileGate = new(false, @"Local\AntiPilot.Log");

    private const long MaxBytes = 256 * 1024;

    public static string LogPath => Path.Combine(AppConfig.ConfigDirectory, "antipilot.log");

    /// <summary>The previous log. Kept so a rotation does not throw away the evidence.</summary>
    public static string PreviousLogPath => LogPath + ".1";

    public static void Write(string message)
    {
        try
        {
            lock (Gate)
            {
                var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff}  {message}{Environment.NewLine}";
                var bytes = System.Text.Encoding.UTF8.GetBytes(line);

                bool held = false;
                try
                {
                    try
                    {
                        // A press that died mid-write leaves the mutex abandoned, which still means
                        // it is ours now. A wait that times out is not worth stalling a key press
                        // over, so the line is simply dropped.
                        held = FileGate.WaitOne(TimeSpan.FromSeconds(2));
                    }
                    catch (AbandonedMutexException)
                    {
                        held = true;
                    }

                    if (!held)
                    {
                        return;
                    }

                    Directory.CreateDirectory(AppConfig.ConfigDirectory);
                    RotateIfLarge();

                    // One Write of the whole line, under the mutex, so no other process can land
                    // between the timestamp and the newline.
                    using var stream = new FileStream(
                        LogPath, FileMode.Append, FileAccess.Write, FileShare.ReadWrite);
                    stream.Write(bytes, 0, bytes.Length);
                }
                finally
                {
                    if (held)
                    {
                        FileGate.ReleaseMutex();
                    }
                }
            }
        }
        catch
        {
            // Logging must never take the app down.
        }
    }

    /// <summary>
    /// Moves the log aside once it gets big rather than deleting it. Whatever explains the problem
    /// being investigated is usually the part that has just scrolled past the size limit.
    /// </summary>
    private static void RotateIfLarge()
    {
        try
        {
            var file = new FileInfo(LogPath);
            if (!file.Exists || file.Length <= MaxBytes)
            {
                return;
            }

            File.Move(LogPath, PreviousLogPath, overwrite: true);
        }
        catch (IOException)
        {
            // Another process has it open. It will be rotated by whoever writes next.
        }
        catch (UnauthorizedAccessException)
        {
            // Same again; not worth failing a log write over.
        }
    }
}

File: src/AntiPilot.Key/Log.cpp
#include "Log.h"

#include "Paths.h"
#include "Text.h"

#include <Windows.h>

#include <format>

namespace AntiPilot::Log
{
    namespace
    {
        /// <summary>Shared with the .NET processes, which write the same file under the same name.</summary>
        constexpr const wchar_t* GateName = L"Local\\AntiPilot.Log";

        constexpr LONGLONG MaxBytes = 256 * 1024;

        /// <summary>
        /// Moves the log aside once it gets big rather than deleting it. Whatever explains the
        /// problem being investigated is usually the part that has just scrolled past the limit.
        /// </summary>
        void RotateIfLarge(const std::wstring& path)
        {
            WIN32_FILE_ATTRIBUTE_DATA info{};
            if (!GetFileAttributesExW(path.c_str(), GetFileExInfoStandard, &info))
            {
                return;
            }

            LARGE_INTEGER size;
            size.HighPart = static_cast<LONG>(info.nFileSizeHigh);
            size.LowPart = info.nFileSizeLow;
            if (size.QuadPart <= MaxBytes)
            {
                return;
            }

            // Another process holding it open makes this fail; whoever writes next rotates it instead.
            MoveFileExW(path.c_str(), PreviousPath().c_str(), MOVEFILE_REPLACE_EXISTING);
        }

        /// <summary>Holds the cross-process gate for the lifetime of one write.</summary>
        class Gate
        {
        public:
            Gate()
            {
                _mutex = CreateMutexW(nullptr, FALSE, GateName);
                if (_mutex == nullptr)
                {
                    return;
                }

                // A press that died mid-write leaves the mutex abandoned, which still means it is
                // ours now. A wait that times out is not worth stalling a key press over.
                DWORD wait = WaitForSingleObject(_mutex, 2000);
                _held = wait == WAIT_OBJECT_0 || wait == WAIT_ABANDONED;
            }

            ~Gate()
            {
                if (_held)
                {
                    ReleaseMutex(_mutex);
                }

                if (_mutex != nullptr)
                {
                    CloseHandle(_mutex);
                }
            }

            Gate(const Gate&) = delete;
            Gate& operator=(const Gate&) = delete;

            bool Held() const { return _held; }

        private:
            HANDLE _mutex = nullptr;
            bool _held = false;
        };
    }

    std::wstring Path()
    {
        return Paths::ConfigDirectory() + L"\\antipilot.log";
    }

    std::wstring PreviousPath()
    {
        return Path() + L".1";
    }

    void Write(std::wstring_view message)
    {
        SYSTEMTIME now;
        GetLocalTime(&now);

        // Same shape as the .NET side's line, so the one file reads as one log.
        std::string line = Text::ToUtf8(std::format(
            L"{:04}-{:02}-{:02} {:02}:{:02}:{:02}.{:03}  {}\r\n",
            now.wYear, now.wMonth, now.wDay, now.wHour, now.wMinute, now.wSecond, now.wMilliseconds,
            message));

        Gate gate;
        if (!gate.Held())
        {
            return;
        }

        std::wstring directory = Paths::ConfigDirectory();
        if (!Paths::EnsureDirectory(directory))
        {
            return;
        }

        std::wstring path = Path();
        RotateIfLarge(path);

        HANDLE file = CreateFileW(
            path.c_str(), FILE_APPEND_DATA, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr,
            OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr);
        if (file == INVALID_HANDLE_VALUE)
        {
            return;
        }

        // One write of the whole line, under the mutex, so no other process can land between the
        // timestamp and the newline.
        DWORD written = 0;
        WriteFile(file, line.data(), static_cast<DWORD>(line.size()), &written, nullptr);
        CloseHandle(file);
    }
}
